### PR TITLE
Add i18n update script

### PR DESCRIPTION
**Pull request by fluiddot**
Follows up https://github.com/wordpress-mobile/gutenberg-mobile/pull/4359 and related to https://github.com/wordpress-mobile/gutenberg-mobile/issues/2162.

**Installable build can be obtained from the following PRs:**
- [Android](https://github.com/wordpress-mobile/WordPress-Android/pull/15778)
- [iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/17705)

## Description

This PR incorporates a new script to handle the process of updating the i18n localizations including improvements, the process consists of the following steps:
1. Extract used strings of plugins (including Gutenberg) from source-map files.
2. Download translations and optimize them for each plugin.
3. Override the i18n cache of `react-native-editor` package, as we're currently doing ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/blob/29fb512d13ebb96381029537ae9f2c595c7b5ed8/bin/i18n-download.sh#L9-L11)).
4. Generate localization strings files to incorporate translations only referenced by native-only JS files.

More information about the scripts involved in the process can be found in the [Gutenberg PR](https://github.com/WordPress/gutenberg/pull/37482).

### Verify i18n update command
1. Execute `npm run i18n:update`.
2. Observe that the content of folder `src/i18n-cache` contains the optimized translations of Gutenberg, Jetpack and Layout Grid.
3. Observe that localization strings files remain unmodified.

**NOTE:** I spotted the following warnings that might require further investigation, however, I think it shouldn't block the PR as we're already missing them and can be addressed a new PR:

**Gutenberg:**
```
WARNING: The following strings are missing from translations:
[
  'Pixels (px)',
  'Percentage (%)',
  'Percent (%)',
  'Relative to parent font size (em)',
  'Relative to root font size (rem)',
  'Viewport width (vw)',
  'Viewport height (vh)',
  'Viewport smallest dimension (vmin)',
  'Viewport largest dimension (vmax)',
  'Width of the zero (0) character (ch)',
  'x-height of the font (ex)',
  'Centimeters (cm)',
  'Millimeters (mm)',
  'Inches (in)',
  'Picas (pc)',
  'Points (pt)',
  'Linear',
  'Radial',
  'Angle',
  'Newest to oldest',
  'Oldest to newest',
  'A → Z',
  'Z → A',
  'Order by',
  'Relative to parent font size (em)\x04ems',
  'Relative to root font size (rem)\x04rems'
]

WARNING: The following native-only strings are already being included in translations:
[ 'Featured' ]
```

**Layout Grid:**
```
WARNING: The following strings are missing from translations:
[ 'One column', 'Two columns', 'Three columns', 'Four columns' ]
```

### ~~Verify that bundle command executes i18n update command~~
~~1. Execute `npm run bundle`
2. Observe that the `i18n:update` command is executed.~~

~~**NOTE:** It's expected to see modifications on files under `bundle/ios/assets/gutenberg/packages/react-native-editor/i18n-cache/data`. Once the PR is ready to merge, we could update the bundle and push these changes.~~

The i18n update command has been unlinked from the bundle generation ([commit reference 1](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/ac35a6ff93e09862e34c55e743e956fcd9b596c8)) because it was making some CI jobs fail ([reference](https://buildkite.com/wordpress-mobile/gutenberg-mobile/builds/1719#34a373d2-b44b-4186-bcfe-219eba666b6c/170-785), [commit reference 2](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/0c54bd690f29e3a65218eb66f5dbbc3ab1b24303)) due to the use of PHP. It shouldn't block the PR as this can be addressed later on.

**NOTE:** This issue will be addressed in a [different PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4411) once the Docker image is updated in https://github.com/wordpress-mobile/docker-gb-mobile-image/pull/3. In the meantime, we'll keep the same behavior as we currently have in `develop` when generating the JS bundle.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/RELEASE-NOTES.txt) if necessary.

### CODE REVIEW

File: src/i18n-cache/index.js
Comment (fluiddot) [inline src/i18n-cache/index.js]: This logic is almost identical to the same file we have in Gutenberg ([reference](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/i18n-cache/index.js)), so it has been deleted and a new file to unify the logic was created [here](https://github.com/WordPress/gutenberg/blob/e47a69fa45793847a7c91a06b9f84ad996db5bef/packages/react-native-editor/bin/i18n-translations-download.js).

File: bin/po2android.js
Comment (fluiddot) [inline bin/po2android.js]: The main change here is that the strings are now extracted from the used strings object (generated by [this script](https://github.com/WordPress/gutenberg/blob/e47a69fa45793847a7c91a06b9f84ad996db5bef/packages/react-native-editor/bin/extract-used-strings.js)) instead of the POT files.

File: bin/po2swift.js
Comment (fluiddot) [inline bin/po2swift.js]: The main change here is that the strings are now extracted from the used strings object (generated by [this script](https://github.com/WordPress/gutenberg/blob/e47a69fa45793847a7c91a06b9f84ad996db5bef/packages/react-native-editor/bin/extract-used-strings.js)) instead of the POT files.

File: bin/po2swift.js
Comment (fluiddot) [inline bin/po2swift.js]: Used strings are grouped by domain, since we only provide one localization strings files, we unify them here and their translations will be requested as part of the main apps' strings.

File: bin/po2swift.js
Comment (fluiddot) [inline bin/po2swift.js]: We only add the localization strings files the strings referenced in `.native.js` and `.ios.js` files.

File: bin/po2swift.js
Comment (fluiddot) [inline bin/po2swift.js]: Most of the changes in this file are produced by the fact that it wasn't properly formatted.

File: bin/po2android.js
Comment (fluiddot) [inline bin/po2android.js]: Most of the changes in this file are produced by the fact that it wasn't properly formatted.

File: bin/po2android.js
Comment (fluiddot) [inline bin/po2android.js]: Used strings are grouped by domain, since we only provide one localization strings files, we unify them here and their translations will be requested as part of the main apps' strings.

File: bin/po2android.js
Comment (fluiddot) [inline bin/po2android.js]: We only add the localization strings files the strings referenced in `.native.js` and `.android.js` files.

File: bin/i18n-download.sh
Comment (fluiddot) [inline bin/i18n-download.sh]: This part is now done in the `i18n-update` script.

File: bin/i18n-update.sh
Comment (fluiddot) [inline bin/i18n-update.sh]: We'll use the new script created in [this PR](https://github.com/WordPress/gutenberg/pull/37482) for downloading translations.

File: bin/i18n-update.sh
Comment (fluiddot) [inline bin/i18n-update.sh]: The translations of Gutenberg are imported from a file in Gutenberg ([reference](https://github.com/WordPress/gutenberg/blob/78c7f0c7a9c3d69979d0633375325104fb03b86d/packages/react-native-editor/src/index.js#L10)), so we overwrite that content with was produced by this script.

File: bin/i18n-update.sh
Comment (fluiddot) [inline bin/i18n-update.sh:76]: By default, the intermediate files produced by the script will be held in a temporary directory.

File: bin/i18n-update.sh
Comment (fluiddot) [inline bin/i18n-update.sh]: Here we validate that the plugin folders provided via the arguments exist before executing the script.

File: bin/i18n-update.sh
Comment (fluiddot) [inline bin/i18n-update.sh]: Similar to the translations download, we'll use the new script created in [this PR](https://github.com/WordPress/gutenberg/pull/37482) for extracting the used strings.

File: bin/i18n-update.sh
Comment (fluiddot) [inline bin/i18n-update.sh]: As mentioned in other comments, now the scripts in charge of generating the localization strings files will use the used strings JSON generated earlier instead of the POT files ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/blob/29fb512d13ebb96381029537ae9f2c595c7b5ed8/package.json#L84-L85)).

File: bin/po2android.js
Comment (fluiddot) [inline bin/po2android.js]: Used strings are grouped by domain, since we only provide one localization strings files, we unify them here and their translations will be requested as part of the main apps' strings.

File: bin/po2swift.js
Comment (fluiddot) [inline bin/po2swift.js]: Used strings are grouped by domain, since we only provide one localization strings files, we unify them here and their translations will be requested as part of the main apps' strings.

File: bin/po2android.js
Comment (fluiddot) [inline bin/po2android.js]: We only add the localization strings files of strings referenced in `.native.js` and `.android.js files`, and that doesn't have context.

File: bin/po2swift.js
Comment (fluiddot) [inline bin/po2swift.js]: We only add the localization strings files of strings referenced in `.native.js` and `.ios.js files`, and that doesn't have context.

File: bin/po2android.js
Comment (fluiddot) [reply bin/po2android.js]: The current list of strings that we won't be able to translate are:
```
[
  'Link rel attribute value placeholder\x04None',
  'inline textinput cell\x04%s. Empty',
  'inline textinput cell\x04%1$s, %2$s. Empty',
  'inline textinput cell\x04%1$s, %2$s',
  'inline textinput cell\x04%1$s, %2$s, %3$s',
  'switch control\x04%s. On',
  'switch control\x04%s. Off',
  'switch control\x04%1$s, %2$s. On',
  'switch control\x04%1$s, %2$s. Off'
]
```

**NOTE:** The strings contain both the value and context, it has the following format: `<CONTEXT>\x04<STRING>`.

File: package.json
Comment (fluiddot) [inline package.json:64]: Eventually, once this script is ready to merge, we could remove all the `makepot` and `wpcli` commands and attach this one to `prebundle:js`.

File: package.json
Comment (fluiddot) [inline package.json]: All these commands are no longer necessary as it's already covered by the i18n update script.

File: package.json
Comment (fluiddot) [inline package.json]: Now when generating a bundle, we'll always update the translations and produce the localization strings files.

File: .buildkite/pipeline.yml
Comment (fluiddot) [inline .buildkite/pipeline.yml]: The command that updates the i18n localizations uses WP-CLI that requires PHP, so we would need to update the [`gb-mobile-image` docker image](https://github.com/wordpress-mobile/docker-gb-mobile-image) and include PHP as part of the installed tools.

File: bin/i18n-update.sh
Comment (jhnstn) [inline bin/i18n-update.sh]: minor nitpick: local variables should be lower case. ALL_UPPERCASE is generally used for exported variables only.

File: bin/i18n-update.sh
Comment (jhnstn) [inline bin/i18n-update.sh]: All these variables should be wrapped in strings to avoid string splitting
see https://github.com/koalaman/shellcheck/wiki/SC2086

File: bin/i18n-update.sh
Comment (jhnstn) [inline bin/i18n-update.sh]: Nitpick: might want to consider using `set -euxo pipefail` see https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/

`set -u` is a bit tricky. just need to use the `INPUT=${VAR:-"some default"}` form to make sure everything is set.

File: bin/i18n-update.sh
Comment (jhnstn) [inline bin/i18n-update.sh:90]: It would be useful to note that the plugin arguments are tuples  i.e. ` domain path/to/plugin` 
That will make it easier to grok why the loop is indexing by 2

File: bin/i18n-update.sh
Comment (jhnstn) [inline bin/i18n-update.sh:76]: Probably should add a trap to clean up the temp file :  
something like `trap 'rm -f "$TARGET_PATH"' EXIT`. 
That will run when the script exits ( well except on `kill -9` )

File: bin/i18n-update.sh
Comment (jhnstn) [inline bin/i18n-update.sh]: It probably doesn't matter here but _might_ want to check the length of `$NOT_FOUND_PLUGIN_FOLDERS` since this `if` is just checking the first element. something like
`if [[ ${#NOT_FOUND_PLUGIN_FOLDERS[@]} -gt 0 ]]; then ...`

**Comment by jhnstn** [comment]
I ran `npm run i18n:update` and got the expected changes in `src/i18n-cache`. I did get slightly different warnings.
for Jetpack : 
```
WARNING: The following native-only strings are already being included in translations:
[ 'Featured', 'Text color' ]
```
I also noticed that `Text color` was added to the android and ios string files.

I should also mention that I ran into a problem with building the TypeScript files (specifically `npx tsc --build` was failing when ran in `./gutenberg`). I had to clean out the node modules in Gutenberg to get past the errors. I'm assuming that is something that would need to be addressed in Gutenberg but figure I'd mention it here if anyone else runs into it.

Overall the changes look good. I'll leave it up to you @fluiddot to pick up the bash suggestions I left. None of them are blockers however. Let me know when you get back if you want merge as is or make those changes.

File: bin/i18n-update.sh
Comment (jhnstn) [reply bin/i18n-update.sh]: It's been awhile since I've read that article 😊  Probably don't want `set -x` in there once this is ready to ship but it helps for debugging.

Comment (fluiddot) [reply bin/i18n-update.sh]: Yep, I wouldn't add the `-x` option as it will print too much debug info.

Comment (fluiddot) [reply bin/i18n-update.sh]: I applied the suggestion in https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/8bc16298d8302cd8052e1cb59333def57759d97a.

File: bin/i18n-update.sh
Comment (fluiddot) [reply bin/i18n-update.sh]: I've renamed all local variables in https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/9274f89558f303ea360545d9e170ba0c53489a20. Additionally, I updated the script `generate-pot-files.sh` in Gutenberg with the same format ([reference](https://github.com/WordPress/gutenberg/pull/37482/commits/f60f993c4476eb5174eb4525f3ae0e7cd7723af3)).

File: bin/i18n-update.sh
Comment (fluiddot) [reply bin/i18n-update.sh]: Good catch, I've applied this in https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/8a7dd7aadd07574c5051f2f5a78aa24dbd438f1d. And similar to local variable naming, I also updated the script `generate-pot-files.sh in Gutenberg ([reference](https://github.com/WordPress/gutenberg/pull/37482/commits/f60f993c4476eb5174eb4525f3ae0e7cd7723af3)).

File: bin/i18n-update.sh
Comment (fluiddot) [reply bin/i18n-update.sh:76]: Oh I didn't know about that command 🤩 , I added this in https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/a9cc79e9051e094d54dd161d2d576c9513311cb3.

> That will run when the script exits ( well except on kill -9 )
I don't think this should be a major issue as it's a temporary folder and eventually it will be deleted by the OS.

File: bin/i18n-update.sh
Comment (fluiddot) [reply bin/i18n-update.sh:90]: Right, I updated the error message in https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/c20f4d6603805f8db499bd723ab8e7e44ca7c2d0.

File: bin/i18n-update.sh
Comment (fluiddot) [reply bin/i18n-update.sh]: As far as I checked both options (non-zero vs array length) work but since the value is an array I agree it's more consistent checking its length 👍 .

I modified it in https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/7edfa4d026c74961768008a670e7f56e6294d9af. Note that due to the use of option `-x` (treat unset variables as an error), the way to obtain the array length is a bit different. I followed the recommendation from [this StackOverflow article](https://stackoverflow.com/a/61551944) related to the safe idiom for this calculation.

**Review by fluiddot — COMMENTED** [review]
Thanks @jhnstn for reviewing the changes 🙇 , I've replied to your comments and applied the suggestions.

Regarding the feedback from https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366#issuecomment-1000542029:

> I did get slightly different warnings.
> for Jetpack :
> 
> WARNING: The following native-only strings are already being included in translations:
> [ 'Featured', 'Text color' ]
> I also noticed that Text color was added to the android and ios string files.

Looks like this string was recently added ([PR reference](https://github.com/WordPress/gutenberg/pull/36028)) so when I generated the strings during testing, I didn't have this change.

It's interesting because this is a case we might run into in the future, as far as I checked, the string `Text color` is included in the Gutenberg GlotPress project ([reference](https://translate.wordpress.org/projects/wp-plugins/gutenberg/dev/en-gb/default/?filters%5Bterm%5D=Text+color&filters%5Bterm_scope%5D=scope_any&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filters%5Bcase_sensitive%5D=yes&filters%5Buser_login%5D=&filter=Apply+Filters&sort%5Bby%5D=priority&sort%5Bhow%5D=desc)) because it was referenced in the web version but the value was changed months ago ([reference](https://github.com/WordPress/gutenberg/pull/34680/files#diff-6ef6519967765b9d4a49b27938bf474f58270c38d8f5358cf4c878c97e2b4276L21-R21)). Not sure why it's still referenced in the GlotPress project but I'd expect that in next Gutenberg plugin releases it's removed, so I think it should be safe to consider the string as part of the native version even though it will be eventually duplicated.

> I should also mention that I ran into a problem with building the TypeScript files (specifically npx tsc --build was failing when ran in ./gutenberg). I had to clean out the node modules in Gutenberg to get past the errors. I'm assuming that is something that would need to be addressed in Gutenberg but figure I'd mention it here if anyone else runs into it.

Sorry to hear you had problems when compiling the Typescript files, did you experience it when running the command `npm run build:gutenberg`?

> Overall the changes look good. I'll leave it up to you @fluiddot to pick up the bash suggestions I left. None of them are blockers however. Let me know when you get back if you want merge as is or make those changes.

Awesome, I already included the suggestions so I'd appreciate it if you could take another look to the changes, thanks 🙇 !

File: .buildkite/pipeline.yml
Comment (fluiddot) [reply .buildkite/pipeline.yml]: This will be solved in https://github.com/wordpress-mobile/docker-gb-mobile-image/pull/3.

**Comment by jhnstn** [comment]
> Sorry to hear you had problems when compiling the Typescript files, did you experience it when running the command `npm run build:gutenberg`?

I believe that was the parent command but it was for sure an issue trying to transpile the Typescript. I didn't trace back the root issue since it was complaining about something in `gutenberg/node_modules`. I would say it's a an unlikely edge case we might not see again. Just wanted to note that reinstalling the node modules fixed the issue.

**Review by jhnstn — APPROVED** [review]
All the changes look good. Excellent work @fluiddot

**Comment by fluiddot** [comment]
@jhnstn thank you very much for the review ❤️  ! I'll mark your last comments as resolved ✅ .

Regarding this statement from the PR's description:
> The i18n update command has been unlinked from the bundle generation ([commit reference 1](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/ac35a6ff93e09862e34c55e743e956fcd9b596c8)) because it was making some CI jobs fail ([reference](https://buildkite.com/wordpress-mobile/gutenberg-mobile/builds/1719#34a373d2-b44b-4186-bcfe-219eba666b6c/170-785), [commit reference 2](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4366/commits/0c54bd690f29e3a65218eb66f5dbbc3ab1b24303)) due to the use of PHP. It shouldn't block the PR as this can be addressed later on.

I realized that the generated bundle for Android actually requires to include the localizations, otherwise, we would miss all translations. Therefore, we need to re-enable the `prebundle:js` command and provide either the original logic related to downloading the translations or use the new i18n update command, however for the latter, the PR would be blocked until the Docker image is updated ([reference](https://github.com/wordpress-mobile/docker-gb-mobile-image/pull/3)).

At this point and in order to prevent blocking the PR, I'm leaning towards providing the original logic and attaching the i18n update command in a new PR. This would imply making some small changes so I'd appreciate it if you could take another look once I push them, thanks 🙇

File: bin/i18n-update.sh
Comment (fluiddot) [inline bin/i18n-update.sh:127]: The previous scripts used here`po2android`/`po2swift` have been renamed to `strings2android`/`strings2swift` which matches the file we use for conversion (used strings JSON files).

**Review by jhnstn — APPROVED** [review]
>  I'm leaning towards providing the original logic and attaching the i18n update command in a new PR.

Sounds good to me, :shipit:

**Pull request merged**
